### PR TITLE
Twl/v1.0.0

### PR DESCRIPTION
**Pull request by Tadeuchi**

### CODE REVIEW

File: src/SqliteContractCache.ts [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+  PruneStats,
+  SortKeyCacheResult,
+  BasicSortKeyCache
+} from "warp-contracts";
+
+import Database from "better-sqlite3";
+import { SqliteCacheOptions } from "./SqliteCacheOptions";
+
+export class SqliteContractCache<V> implements BasicSortKeyCache<V> {
+  private readonly logger = LoggerFactory.INST.create("SQLite contract cache");
```
Comment (ppedziwiatr) [inline src/SqliteContractCache.ts]: SqliteContractCache.name

File: src/SqliteContractCache.ts [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+  // Lazy initialization upon first access
+  private get db() {
+    if (!this._db) {
+      if (this.cacheOptions.inMemory) {
+        this._db = new Database(":memory:");
+      } else {
+        if (!this.cacheOptions.dbLocation) {
+          throw new Error(
+            "LevelDb cache configuration error - no db location specified"
```
Comment (ppedziwiatr) [inline src/SqliteContractCache.ts]: not leveldb

File: src/SqliteContractCache.ts [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    }
+    return this._db;
+  }
+
+  private sortKeyTable() {
+    this.db.exec(`
+        CREATE TABLE IF NOT EXISTS sort_key_cache
+        (
+            id
+            INTEGER
```
Comment (ppedziwiatr) [inline src/SqliteContractCache.ts]: chujowo się czyta z takim formatowaniem :-)

File: src/SqliteContractCache.ts [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            PRIMARY
+            KEY,
+            key
```
Comment (ppedziwiatr) [inline src/SqliteContractCache.ts]: dont we need to add some indexes here?

Comment (ppedziwiatr) [reply src/SqliteContractCache.ts]: oh sorry, uniqie (key, sort_key).....formatting....

File: src/SqliteContractCache.ts [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+      .prepare(
+        `SELECT count(*)
+       FROM sort_key_cache`
+      )
+      .pluck()
+      .get();
+    const result = this.db
+      .prepare(
+        `
+            WITH sorted_cache AS
```
Comment (ppedziwiatr) [inline src/SqliteContractCache.ts:228]: and this works per contract, right? i.e. partition by "key" guarantees that, right?

File: src/SqliteContractCache.ts [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+      return new SortKeyCacheResult<V>(
+        result.sort_key,
+        JSON.parse(result.value)
+      );
+    }
+    return null;
+  }
+
+  async put(stateCacheKey: CacheKey, value: V): Promise<void> {
+    const strVal = JSON.stringify(value);
```
Comment (ppedziwiatr) [inline src/SqliteContractCache.ts]: maybe we should use safe-stable-strinigify here, to get stable and detereministic results?

File: src/SqliteContractCache.ts [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    this.db
+      .prepare(
+        `
+            WITH sorted_cache AS
+                     (SELECT id, row_number() over (ORDER BY sort_key DESC) AS rw
+                      FROM sort_key_cache
+                      WHERE key = ?)
+            DELETE
+            FROM sort_key_cache
+            WHERE id IN (SELECT id FROM sorted_cache WHERE rw > ?);
```
Comment (ppedziwiatr) [inline src/SqliteContractCache.ts:153]: what about VACUUUUUUUUUUUUUM? I guess sqlite does not run vacuum on itself?

File: src/SqliteContractCache.ts (hunk 5 above)
Comment (Tadeuchi) [reply src/SqliteContractCache.ts:228]: yes, this query will group items by key (in this case contract id), order by sortKey desc and assign row_number

File: src/SqliteContractCache.ts (hunk 7 above)
Comment (Tadeuchi) [reply src/SqliteContractCache.ts:153]: yes, you are right, VACUUUUUUUUUUUUUM is pretty important here if don't want to end up with a HUUUGE file
I tried some aprroaches such as 
- counting deleted items and calling vaccum after some threshold
It works, but..
Maybe we should try Incremental auto-vacuum?
It reclaims space marked as deleted.

File: src/SqliteContractCache.ts (hunk 1 above)
Comment (Tadeuchi) [reply src/SqliteContractCache.ts]: done

File: src/SqliteContractCache.ts (hunk 2 above)
Comment (Tadeuchi) [reply src/SqliteContractCache.ts]: done

File: src/SqliteContractCache.ts (hunk 3 above)
Comment (Tadeuchi) [reply src/SqliteContractCache.ts]: fixed

File: src/SqliteContractCache.ts (hunk 6 above)
Comment (Tadeuchi) [reply src/SqliteContractCache.ts]: yes, definitely, changed

**Pull request merged**